File: corpus/ops/cli/src/ofctlpkg/commands/telemetry.nim
##[
  أوامر التنبيهات وقراءات المستشعرات عند telemetry-ingest: ما يفتحه المناوب أولاً حين يرنّ
  إنذار سلسلة التبريد. يعرض التنبيهات المفتوحة، يقرّها، يغلقها بعد معالجتها ميدانياً، ويسحب
  نافذة قراءات من قسم الإقليم كي يرى المنحنى الذي تسبّب في التنبيه.
]##

import std/[json, strutils, strformat, parseopt, times]
import ../[context, apiclient, render]

const
  # قائمة CHECK على `telemetry.telemetry_alerts.rule_code` كما في §2.5.
  RuleCodes = [
    "temp_excursion_high", "temp_excursion_low", "humidity_high", "shock_impact",
    "door_open_in_transit", "battery_critical", "gateway_silent", "geofence_breach"
  ]

proc listAlerts(ctx: var OpsContext, state, ruleCode, containerId: string, severityMin: int) =
  ## `GET /v1/alerts` مع مرشّحات §3.4. تذكّر أن هذه القائمة هي منظور telemetry-ingest وحده:
  ## قلب الشحنة إلى `at_risk` يتمّ عند container-registry بعد استهلاك telemetry.alert.raised
  ## (§4.9) وبشرط أن تبلغ الخطورة OF_FREIGHT_AUTO_AT_RISK_SEVERITY، فوجود تنبيه خطورته 2
  ## وشحنة ما زالت `in_transit` ليس تناقضاً بل هو السلوك المضبوط.
  if ruleCode.len > 0 and ruleCode notin RuleCodes:
    raise newException(ValueError, "rule_code غير معروف: " & ruleCode & " — راجع §2.5")
  var rows: seq[seq[string]]
  var openCount = 0
  for alert in ctx.paginate("telemetry-ingest", "/v1/alerts", @[
      ("state", state), ("rule_code", ruleCode), ("container_id", containerId),
      ("severity_min", if severityMin > 0: $severityMin else: "")]):
    let closed = alert{"closed_at"}.getStr("")
    if closed.len == 0: inc openCount
    rows.add @[
      alert["alert_id"].getStr(),
      alert["container_id"].getStr(),
      alert{"shipment_id"}.getStr("—"),      # يُحلّ عبر freight.v1.ContainerLookup/ResolveShipmentForContainer وقت الرفع
      alert["rule_code"].getStr(),
      $alert["severity"].getInt(),
      alert["threshold_value"].getStr(),
      alert{"peak_value"}.getStr("—"),
      alert["opened_at"].getStr(),
      if closed.len > 0: "closed" else: (if alert{"acknowledged_at"}.getStr("").len > 0: "acked" else: "open")
    ]
  emitTable(ctx, @["alert_id", "container", "shipment", "rule", "sev",
                   "threshold", "peak", "opened_at", "state"], rows)
  if ctx.format == ofTable and openCount > 0:
    warn(ctx, &"{openCount} تنبيهاً مفتوحاً؛ التنبيه المفتوح يبقي الشحنة في at_risk ويمنع تسليمها.")

proc acknowledgeAlert(ctx: var OpsContext, alertId: string) =
  ## الإقرار يوقف تكرار الإشعار عبر notification-service ولا يغلق التنبيه: يبقى
  ## `closed_at` فارغاً وتبقى الشحنة في `at_risk` حتى يُحلّ السبب فعلاً على الأرض.
  let doc = ctx.request("telemetry-ingest", &"/v1/alerts/{alertId}/acknowledge",
                        httpMethod = HttpPost, body = %*{})
  if ctx.dryRun: return
  echo &"{alertId}: أُقرّ بواسطة {doc[\"acknowledged_by\"].getStr()} في {doc[\"acknowledged_at\"].getStr()}"

proc closeAlert(ctx: var OpsContext, alertId, note: string) =
  ## الإغلاق اليدوي مخصّص للحالة التي عولجت ميدانياً بينما ما زال المستشعر يقرأ قيمة سيّئة
  ## (باب مفتوح أثناء التفريغ مثلاً). لا يمحو التنبيه: الصف يبقى في `telemetry.telemetry_alerts`
  ## ونسخته التي لا تُعدَّل موجودة في `platform.audit_ledger_entries`.
  if note.len == 0:
    raise newException(ValueError, "--note مطلوب عند الإغلاق اليدوي؛ يُقرأ في مراجعة الحادثة")
  let doc = ctx.request("telemetry-ingest", &"/v1/alerts/{alertId}/close",
                        httpMethod = HttpPost, body = %*{"note": note})
  if ctx.dryRun: return
  echo &"{alertId}: أُغلق في {doc[\"closed_at\"].getStr()}"

proc showReadings(ctx: var OpsContext, containerId: string, sinceMinutes: int) =
  ## `GET /v1/containers/{container_id}/readings` استعلام نافذة زمنية يصيب قسم الإقليم وحده
  ## (`telemetry.telemetry_readings` مقسّمة LIST على region_code). طلب حاوية خارج إقليمنا يعود
  ## بـ 403 لا بإعادة توجيه، وهذا مقصود: OF_TELEMETRY_ALLOWED_REGIONS يمنع الخلط لا يصلحه.
  let since = (getTime() - initDuration(minutes = sinceMinutes)).utc.format("yyyy-MM-dd'T'HH:mm:ss'Z'")
  var rows: seq[seq[string]]
  for r in ctx.paginate("telemetry-ingest", &"/v1/containers/{containerId}/readings",
                        @[("since", since)]):
    rows.add @[
      r["reading_id"].getStr(),
      r["recorded_at"].getStr(),
      r["received_at"].getStr(),
      r{"temperature_c"}.getStr("—"),
      r{"humidity_pct"}.getStr("—"),
      r{"shock_g"}.getStr("—"),
      if r{"door_open"}.getBool(false): "open" else: "closed",
      $r{"battery_pct"}.getInt(0),
      r["gateway_id"].getStr()
    ]
  emitTable(ctx, @["reading_id", "recorded_at", "received_at", "temp_c",
                   "humidity", "shock_g", "door", "battery", "gateway"], rows)
  if rows.len == 0:
    warn(ctx, "لا قراءات في النافذة المطلوبة — تحقّق من صمت البوابة عبر " &
              "alert list --rule=gateway_silent قبل أن تشكّ في الحاوية نفسها.")

proc run*(ctx: var OpsContext, command: string, args: seq[string]): int =
  var positional: seq[string]
  var state = "open"
  var ruleCode, containerId, note = ""
  var severityMin = 0
  var sinceMinutes = 120

  var parser = initOptParser(args)
  for kind, key, val in parser.getopt():
    case kind
    of cmdArgument: positional.add key
    of cmdLongOption, cmdShortOption:
      case key
      of "state":        state = val
      of "rule":         ruleCode = val
      of "container":    containerId = requirePrefix(val, "cnt_")
      of "severity-min": severityMin = parseInt(val)
      of "since-minutes": sinceMinutes = parseInt(val)
      of "note":         note = val
      else: discard
    of cmdEnd: discard

  case command
  of "list":
    ctx.listAlerts(state, ruleCode, containerId, severityMin)
  of "ack":
    ctx.acknowledgeAlert(requirePrefix(positional[2], "alr_"))
  of "close":
    ctx.closeAlert(requirePrefix(positional[2], "alr_"), note)
  of "readings":
    ctx.showReadings(requirePrefix(positional[2], "cnt_"), sinceMinutes)
  else:
    stderr.writeLine "أوامر alert: list | ack | close | readings"
    return 64
  0

File: corpus/web/src/api/endpoints/telemetry.ts
/*
 * telemetry-ingest 的只读面：告警列表、认领与关闭，以及冷链曲线要用的读数窗口查询。
 * 控制台从不往 /v1/ingest/batch 写东西 —— 那条路径只属于 edge/ 下的网关代理，
 * 而且要求 Ed25519 签名，浏览器既没有私钥也不该有。
 */

import { fetchPage, type Page, type PageRequest } from '../pagination';
import type { ApiClient, RequestOptions } from '../http-client';
import type { AlertRuleCode, AlertSeverity, LatLon, RegionCode } from '../../types/domain';
import type { AlertId, ContainerId, GatewayId, ReadingId, ShipmentId, UserId } from '../../types/ids';

const SERVICE = 'telemetry-ingest' as const;

/** `GET /v1/alerts` 的一行，对应 `telemetry.telemetry_alerts`。 */
export interface TelemetryAlert {
  readonly alert_id: AlertId;
  readonly container_id: ContainerId;
  /** 起警时通过 freight.v1.ContainerLookup/ResolveShipmentForContainer 解析出来；空箱时为 null。 */
  readonly shipment_id: ShipmentId | null;
  readonly rule_code: AlertRuleCode;
  readonly severity: AlertSeverity;
  readonly opened_at: string;
  readonly closed_at: string | null;
  readonly peak_value: number | null;
  readonly threshold_value: number;
  readonly first_reading_id: ReadingId;
  readonly acknowledged_by: UserId | null;
  readonly acknowledged_at: string | null;
}

/** 一条读数。冷链曲线只画 `temperature_c`，其余字段给详情浮层用。 */
export interface TelemetryReading {
  readonly reading_id: ReadingId;
  readonly region_code: RegionCode;
  readonly container_id: ContainerId;
  readonly gateway_id: GatewayId;
  /** 传感器自己的时钟。 */
  readonly recorded_at: string;
  /** telemetry-ingest 收到的时刻。断链补传时两者可以差好几个小时。 */
  readonly received_at: string;
  readonly temperature_c: number | null;
  readonly humidity_pct: number | null;
  readonly shock_g: number | null;
  readonly door_open: boolean | null;
  readonly battery_pct: number | null;
  readonly position: LatLon | null;
}

/**
 * 告警列表。
 *
 * `severity_min` 通常设成与 container-registry 的 `OF_FREIGHT_AUTO_AT_RISK_SEVERITY`
 * 相同的值，这样值班面板上看到的就正好是「会把运单打成 at_risk 的那一批」。
 */
export function listAlerts(
  client: ApiClient,
  filters: {
    state?: 'open' | 'acknowledged' | 'closed';
    rule_code?: AlertRuleCode;
    container_id?: ContainerId;
    severity_min?: AlertSeverity;
  },
  request?: PageRequest,
  options?: RequestOptions,
): Promise<Page<TelemetryAlert>> {
  return fetchPage<TelemetryAlert>(
    client,
    SERVICE,
    '/v1/alerts',
    { ...request, filters: { ...request?.filters, ...filters } },
    options,
  );
}

/**
 * 认领一条告警。只写 `acknowledged_by` / `acknowledged_at`，不关闭 ——
 * 认领的语义是「我在看了」，而现场温度是不是真的回来了，认领的人当时并不知道。
 */
export function acknowledgeAlert(
  client: ApiClient,
  alertId: AlertId,
  options?: RequestOptions,
): Promise<TelemetryAlert> {
  return client.post<TelemetryAlert>(SERVICE, `/v1/alerts/${alertId}/acknowledge`, options);
}

/**
 * 手工关闭。规则自己不会关某些告警（比如 `shock_impact` 是一次性事件，
 * 不存在「恢复正常」的读数），所以现场处理完之后必须有人来点这个。
 */
export function closeAlert(
  client: ApiClient,
  alertId: AlertId,
  resolutionNote: string,
  options?: RequestOptions,
): Promise<TelemetryAlert> {
  return client.post<TelemetryAlert>(SERVICE, `/v1/alerts/${alertId}/close`, {
    ...options,
    body: { resolution_note: resolutionNote },
  });
}

/**
 * 时间窗读数查询，直接打到该区域的分区上。
 *
 * 跨区域是查不到的：`telemetry.telemetry_readings` 按 region_code 做 LIST 分区，
 * 一条巴西的读数不会出现在欧洲的分区里，这是数据驻留要求而不是性能优化（SPEC §7 规则 7）。
 * 所以窗口再宽，也只能看到当前区域集群里的数据。
 */
export function listReadings(
  client: ApiClient,
  containerId: ContainerId,
  window: { from: string; to: string },
  request?: PageRequest,
  options?: RequestOptions,
): Promise<Page<TelemetryReading>> {
  return fetchPage<TelemetryReading>(
    client,
    SERVICE,
    `/v1/containers/${containerId}/readings`,
    { ...request, filters: { ...request?.filters, from: window.from, to: window.to } },
    options,
  );
}

/**
 * 把读数抽稀到图表宽度，避免把七天、每分钟一条的曲线一次性塞进 canvas。
 * 保留每个桶里的极值点，因为冷链看的是越界峰值，平均值会把 excursion 抹平。
 */
export function downsamplePeaks(
  readings: ReadonlyArray<TelemetryReading>,
  buckets: number,
): TelemetryReading[] {
  if (readings.length <= buckets || buckets <= 0) return [...readings];
  const size = Math.ceil(readings.length / buckets);
  const out: TelemetryReading[] = [];

  for (let i = 0; i < readings.length; i += size) {
    const slice = readings.slice(i, i + size);
    const peak = slice.reduce((best, current) => {
      const a = Math.abs(current.temperature_c ?? 0);
      const b = Math.abs(best.temperature_c ?? 0);
      return a > b ? current : best;
    }, slice[0]);
    out.push(peak);
  }
  return out;
}
